field/TinyLang/Field/Generator.hs

```diff
@@ -420,42 +506,23 @@ instance (KnownUni f a, Field f, Arbitrary f) => Arbitrary (Expr f a) where
         EConst _ -> []
         EVar _ -> []
-        -- TODO: we can safely drop an assertion and we can drop a let-expression when
-        -- the let-bound variable is not used in @expr@.
-        EStatement stat expr -> uncurry EStatement <$> shrink (stat, expr)
-
--- An instance that QuickCheck can use for tests.
-instance (Field f, Arbitrary f) => Arbitrary (SomeUniExpr f) where
-    arbitrary = withOneofUnis $ \uni -> SomeOf uni <$> arbitrary
-
-    shrink (SomeOf uni0 expr) =
-        map (SomeOf uni0) (withKnownUni uni0 $ shrink expr) ++ case expr of
-            EAppUnOp op e -> withUnOpUnis op $ \argUni _ -> [SomeOf argUni e]
-            EAppBinOp op e1 e2 ->
-                withBinOpUnis op $ \uni1 uni2 _ ->
-                    [SomeOf uni1 e1, SomeOf uni2 e2]
-            EIf e _ _ -> [SomeOf Bool e]
-            EConst _ -> []
-            EVar _ -> []
-            EStatement stat _ -> case stat of
-                ELet (UniVar uni _) def -> [SomeOf uni def]
-                EAssert e               -> [SomeOf Bool e]
-
-genEnvFromVarSigns :: (Field f, Arbitrary f) => Env (VarSign f) -> Gen (Env (SomeUniConst f))
-genEnvFromVarSigns =
-    traverse $ \(VarSign _ (uni :: Uni f a)) ->
+
+
+genEnvFromVarSigs :: (Field f, Arbitrary f) => Env (VarSig f) -> Gen (Env (SomeUniConst f))
+genEnvFromVarSigs =
+    traverse $ \(VarSig _ (uni :: Uni f a)) ->
         Some <$> withKnownUni uni (arbitrary :: Gen (UniConst f a))
 
--- | Generate a random ExprWithEnv.  Note that you can say things like
--- "generate (resize 1000 arbitrary :: Gen (ExprWithEnv F17))" to get
--- bigger expressions.  There's no means provided to generate things
--- over non-default sets of variables, but this would be easy to do.
-instance (Field f, Arbitrary f) => Arbitrary (ExprWithEnv f) where
+-- | Generate a random ProgramWithEnv.  Note that you can say things like
+-- "generate (resize 1000 arbitrary :: Gen (ProgramWithEnv F17))" to get bigger
+-- expressions. There's no means provided to generate things over non-default
+-- sets of variables, but this would be easy to do.
+instance (Field f, Arbitrary f) => Arbitrary (ProgramWithEnv f) where
     arbitrary = do
-        someUniExpr <- arbitrary
-        vals <- forget (genEnvFromVarSigns . exprFreeVarSigns) someUniExpr
-        return $ ExprWithEnv someUniExpr vals
-    shrink (ExprWithEnv someUniExpr (Env vals)) =
-        -- TODO: test me.
-        flip map (shrink someUniExpr) $ \shrunk@(SomeOf _ expr) ->
-            ExprWithEnv shrunk . Env . IntMap.intersection vals . unEnv $ exprFreeVarSigns expr
+        prog <- arbitrary
+        vals <- genEnvFromVarSigs . progFreeVarSigs $ prog
+        return $ ProgramWithEnv prog vals
+    shrink (ProgramWithEnv prog (Env vals)) =
+        -- TODO:  Check why is this keeping extra variable bindings in the env
+        flip map (shrink prog) $ \shrunk ->
+            ProgramWithEnv shrunk . Env . IntMap.intersection vals . unEnv $ progFreeVarSigs prog
```

Comment:
Here's one mistake. You need to take `progFreeVarSigs` of the shrunk program, not the original one. Previously it was `exprFreeVarSigns expr` where `expr` comes from `shrunk`.

Reply:
Yep, if fixed, then the shrunk test case looks correctly.

And if we define `shrink` as `const []`, then we get

```
        INITIAL STATE
          Env {unEnv = fromList [(1,Some 2),(7,Some 3),(9,Some (UniConst Bool False)),(23,Some 0)]}
        
        ORIGINAL PROGRAM
          assert (16 / t_7) >= (16 / t_7);
          let q_23 = if ?b_9 then y_1 else q_23;
``` 

which correctly generates values only for free variables.

Reply:
Thanks!

Reply:
Fixed.